src/lib/pipeline/m2/index.js

```diff
@@ -128,99 +251,120 @@ class M2 extends THREE.Group {
     // Mirror geometry over X and Y axes and rotate
     const matrix = new THREE.Matrix4();
     matrix.makeScale(-1, -1, 1);
-    sharedGeometry.applyMatrix(matrix);
-    sharedGeometry.rotateX(-Math.PI / 2);
+    geometry.applyMatrix(matrix);
+    geometry.rotateX(-Math.PI / 2);
 
-    const { textureLookups, textures, renderFlags } = data;
-    const { transparencyLookups, transparencies, colors } = data;
-    const { indices, textureUnits, triangles } = skinData;
+    // Preserve the geometry
+    this.geometry = geometry;
+  }
 
-    // TODO: Look up colors, render flags and what not
-    for (let tuIndex = 0, len = textureUnits.length; tuIndex < len; ++tuIndex) {
-      const textureUnit = textureUnits[tuIndex];
+  createMesh(geometry, skeleton, rootBones) {
+    let mesh;
 
-      const textureLookup = textureLookups[textureUnit.textureIndex];
-      const texture = textures[textureLookup];
-      textureUnit.texture = texture;
+    if (this.useSkinning) {
+      mesh = new THREE.SkinnedMesh(geometry);
 
-      textureUnit.renderFlags = renderFlags[textureUnit.renderFlagsIndex];
+      // Assign root bones to mesh
+      rootBones.forEach((bone) => {
+        mesh.add(bone);
+        bone.skin = mesh;
+      });
 
-      if (textureUnit.transparencyIndex > -1) {
-        const transparencyLookup = transparencyLookups[textureUnit.transparencyIndex];
-        const transparency = transparencies[transparencyLookup];
-        textureUnit.transparency = transparency;
-      }
+      // Bind mesh to skeleton
+      mesh.bind(skeleton);
+    } else {
+      mesh = new THREE.Mesh(geometry);
+    }
 
-      if (textureUnit.colorIndex > -1) {
-        const color = colors[textureUnit.colorIndex];
-        textureUnit.color = color;
-      }
+    // Add mesh to the group
+    this.add(mesh);
+
+    // Assign as root mesh
+    this.mesh = mesh;
+  }
+
+  createSubmeshes(data, skinData) {
+    const { vertices } = data;
+    const { submeshes, indices, triangles } = skinData;
+
+    const subLen = submeshes.length;
+
+    for (let submeshIndex = 0; submeshIndex < subLen; ++submeshIndex) {
+      const submeshDef = submeshes[submeshIndex];
+
+      // Bring up relevant texture units and geometry.
+      const submeshTextureUnits = this.textureUnits.get(submeshIndex);
+      const submeshGeometry = this.submeshGeometries.get(submeshIndex) ||
+        this.createSubmeshGeometry(submeshDef, indices, triangles, vertices);
+
+      const submesh = this.createSubmesh(submeshDef, submeshGeometry, submeshTextureUnits);
+
+      this.parts.set(submesh.userData.partID, submesh);
+      this.submeshes.push(submesh);
+
+      this.submeshGeometries.set(submeshIndex, submeshGeometry);
+
+      this.mesh.add(submesh);
     }
+  }
 
-    const { submeshes } = this.skinData;
+  createSubmeshGeometry(submeshDef, indices, triangles, vertices) {
+    const geometry = this.geometry.clone();
```

Comment:
Look above, at lines 297-300.

I'm obtaining the submesh geometry out of ```this.submeshGeometries``` if it exists. If it doesn't, I create it. This allows for the first construction of a given M2 to produce the submesh geometries, and all subsequent (instanced) M2s to recycle them.